akiraux/Akira
# Linear and Radial Gradients feature on CanvasItems

## Summary / How this PR fixes the problem?
### UI Changes
1. The ColorPicker shows 3 buttons on top for **Solid Color**, **Linear Gradient** and **Radial Gradient**. The buttons show icons based on their coloring mode.
2. Below these buttons, a ```Gtk.EventBox``` is used to place stop colors on the gradients. It only becomes active in **Linear Gradient** or **Radial Gradient** mode.
3. The stop colors can be moved around by clicking on them. Moving the stop colors changes the background color of ```EventBox``` and the selected CanvasItem instantaneously. 
4. StopColors can be deleted by selecting them and pressing the delete button beside the ```Gtk.EventBox```.
5. In order to change the direction of the gradient, 2 nobs are present inside the selected canvas item. 
6. Dragging these nobs changes the gradient direction and automatically updates the shading.

### Code Changes
1. Added icons for ColorMode buttons.
2. Created ColorMode class as a container for Color Mode buttons and the EventBox.
3. Created GradientEditor class for adding new StopColors and modifying their colors.(The class for StopColor is implemented inside GradientEditor).
4. Created DirectionLine class for changing the direction of gradients. Uses circular nobs with button events to drag them around.

## Steps to Test
1. Create a CanvasItem which has a Fill object(like Rectangle or Ellipse)
2. Press the ColorButton in the Fill object. This opens a popover.
3. Click on either of the gradient buttons(2nd for Linear and 3rd for Radial).
4. In the rectangular box that appears below the buttons, add new stop colors by clicking on the horizontal line.
5. With a stop color selected(selected stop color is blue in color), change the color using the color picker below.
6. Click and drag the stop color to change their position.
7. Select some stop color and press the delete button to remove it. (Note that the first and last stop colors cannot be deleted).
8. After you are done playing with the stop colors, click somewhere on the screen to remove the popover.
9. By default, the nobs for gradient direction are present on the diagonal of the canvas item. Drag them to change the direction of the gradient.

## Screenshots 


![stop_color_shot](https://user-images.githubusercontent.com/64722310/124347454-9010d680-dc02-11eb-9a27-3ae6f5aed7a8.png)
![direction_line shot](https://user-images.githubusercontent.com/64722310/124347462-969f4e00-dc02-11eb-9349-fff092dda707.png)


## Known Issues / Things To Do
1. Gradients haven't been implemented for Borders.
2. Minor bug-
    Select an item, then apply gradient to it and change the direction,
    Then select the item again and select gradient. The direction resets to its initial position.
    Fix: Use the gradient pattern from ColorModel to figure out the initial positions of gradient.
3. Gradients cannot be blended with gradients or colors.
4. Major bug- The app crashes when you do the following
    Select an item and apply gradient to it.
    Delete the fill item and create a new fill item.
    Selecting gradient mode in this fill item crashes the app.

## This PR fixes/implements the following **bugs/features**:
 

- Fixes #605
